### PR TITLE
Port e2e push test

### LINKED ISSUE #3837: Port e2e push tests
Port push tests https://github.com/sylabs/singularity/blob/master/e2e/push/push.go to use singularity execution context (https://github.com/sylabs/singularity/blob/master/e2e/internal/e2e/singularitycmd.go)

### PR DESCRIPTION
**Pull request by gvallee**
**Description of the Pull Request (PR):**

Port e2e push test to new API

**This fixes or addresses the following GitHub issues:**

- Fixes #3837 
- Addresses #3828

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: e2e/push/push.go [hunk 1]
```diff
@@ -39,7 +40,7 @@ func (c *ctx) testPushCmd(t *testing.T) {
 		desc          string // case description
 		dstURI        string // destination URI for image
 		imagePath     string // src image path
-		expectSuccess bool   // singularity should exit with code 0
+		expectSuccess bool   // expection regarding the test's success
```
Comment (gvallee) [inline e2e/push/push.go]:
```suggestion
		expectSuccess bool   // expectation regarding the test's success
```

File: e2e/push/push.go [hunk 2]
```diff
@@ -39,7 +40,7 @@ func (c *ctx) testPushCmd(t *testing.T) {
 		desc          string // case description
 		dstURI        string // destination URI for image
 		imagePath     string // src image path
-		expectSuccess bool   // singularity should exit with code 0
+		expectSuccess bool   // expectation regarding the test's success
```
Comment (jscook2345) [inline e2e/push/push.go]: In other tests I've seen this field changed to something like`expectExitCode` and than using that value instead of doing further checking around the `boolean` later on.

What do you think?

File: e2e/push/push.go [hunk 3]
```diff
@@ -75,26 +76,24 @@ func (c *ctx) testPushCmd(t *testing.T) {
...
-
-			case tt.expectSuccess && err != nil:
-				// FAIL: expecting success, failed
-
-				t.Logf("Running command:\n%s\nOutput:\n%s\n", cmd, out)
-				t.Errorf("unexpected failure: %v", err)
-
-			case !tt.expectSuccess && err == nil:
-				// FAIL: expecting failure, succeeded
+			args := tt.dstURI
```
Comment (ikaneshiro) [inline e2e/push/push.go:83]: Can we make `args` a `[]string` and append to it so we dont need to split a string by a separator in `e2e.WithArgs()`?

**Review by ikaneshiro — CHANGES_REQUESTED** [review]

File: e2e/push/push.go (hunk 3 above)
Comment (gvallee) [reply e2e/push/push.go:83]: I had that question before and no it does not work. In a nutshell, the command is very sensitive to empty slots, meaning an element in the slice that would not be defined, which would make the command to fail even if everything looks fine (only a double space would appear when displaying the elements of the commands). I had very bad luck with unit tests, e2e tests and in the context of my MPI tool.
In other words, that is by far the safer way to deal with it and i did not manage to have another solution that never create a problem.

Comment (gvallee) [reply e2e/push/push.go:83]: I had that question before and the short answer is no: so far, i always end up with Go assuming that an element in the slice is empty, which will make the command fail (when displaying the args, the problem can be seen through a double space). So this is the only reliable way i know of dealing with it.
I had the problem in unit tests, e2e tests and the MPI tool.

File: e2e/push/push.go (hunk 2 above)
Comment (gvallee) [reply e2e/push/push.go]: Done

File: e2e/push/push.go (hunk 3 above)
Comment (ikaneshiro) [reply e2e/push/push.go:83]: Could you throw a comment on the code summarizing these comments so that others hopefully dont run into this issue when working with this file?

Comment (jscook2345) [reply e2e/push/push.go:83]: Not handling an empty argument seems like a short coming. @gvallee Can you open an issue for that so we can make things more consistent across the board?

**Review by jscook2345 — APPROVED** [review]

File: e2e/push/push.go (hunk 3 above)
Comment (gvallee) [reply e2e/push/push.go:83]: Will do and as you can see I am having problems with Github, my comments appearing/disappearing from my view of the PR... :(

**Comment by jscook2345** [comment]
@gvallee 

See @mem 's changes here: https://github.com/sylabs/singularity/pull/3840

They will need to be applied before the tests will pass.

**Comment by gvallee** [comment]
I do not understand the API modification introduced in #3840, it will take me some time to figure it out.

File: e2e/push/push.go [hunk 4]
```diff
@@ -75,26 +76,23 @@ func (c *ctx) testPushCmd(t *testing.T) {
...
+			// an empty elements to the list when passing it to the command, which
+			// will create a failure.
+			args := tt.dstURI
+			if tt.imagePath != "" {
+				args = tt.imagePath + " " + args
 			}
+
+			e2e.RunSingularity(
+				t,
+				tt.desc,
```
Comment (mem) [inline e2e/push/push.go]: replace this with `e2e.AsSubtest(tt.desc)`
```suggestion
				e2e.AsSubtest(tt.desc),
```

**Review by mem — CHANGES_REQUESTED** [review]

File: e2e/push/push.go [hunk 5]
```diff
@@ -75,26 +76,23 @@ func (c *ctx) testPushCmd(t *testing.T) {
...
+			// We create the list of arguments using a string instead of a slice of
+			// strings because using slices of strings most of the type ends up adding
+			// an empty elements to the list when passing it to the command, which
+			// will create a failure.
+			args := tt.dstURI
+			if tt.imagePath != "" {
+				args = tt.imagePath + " " + args
 			}
+
+			e2e.RunSingularity(
```
Comment (mem) [inline e2e/push/push.go]: if you change this to `c.env.RunSingularity`, what I wrote below actually works :-|

```suggestion
			c.env.RunSingularity(
```

File: e2e/push/push.go (hunk 4 above)
Comment (gvallee) [reply e2e/push/push.go]: That is the part I do not understand, semantically, I guess I do not understand the meaning of `AsSubtest`, which, without spending more than 2 minutes on it, would infer that I want to create a subtest... which is not the case. But I will look it up, I just need time. :)

Comment (mem) [reply e2e/push/push.go]: Welcome to my world :-)

The function `e2e.RunSingularity` previously had a signature that was something like `func(t *testing.T, name string, ...)`.

I changed it to `func(t *testing.T, cmdPath string, ...)`

In the previous signature, `name` is the name of the subtest (as in `t.Run(name, f)`). If name was empty (`""`), then it _did not_ run a subtest. If name was not empty, it run the command as a subtest.

To make that less awkward and less magic, I changed it to the explicit `e2e.AsSubtest(name)`, which is passed along with all the other options.

But then I added the `cmdPath` argument :-(

The `cmdPath` argument specifies the path to singularity.

I guess I could add a `type CmdPath string` to make this a compiler error (at the cost of making the code a little weird).

Comment (mem) [reply e2e/push/push.go]: And if you don't want a subtest, just don't pass e2e.AsSubtest. I suggested that to match the behavior of the previous code.

Comment (gvallee) [reply e2e/push/push.go]: Thanks for the explanation, it truly helps. I like the fact that we can specify the Singularity binary and now that I have your explanation, I understand the subtest part too. I am totally okay with the new API. :)

**Comment by jscook2345** [comment]
@gvallee @mem 

The PR that keeps on giving ;)

I built locally using this branch and this is what `make check` gave me on `macos`.

```
grapplehammer:singularity justin$ make -C builddir check
 CHECK golangci-lint
cmd/plugin_manifestgen/main.go:32:23: `marshalling` is a misspelling of `marshaling` (misspell)
                sylog.Fatalf("While marshalling manifest to json: %s", err)
```

**Comment by gvallee** [comment]
This is not related to this PR at all but I will fix it as soon as I can.

**Review by ikaneshiro — APPROVED** [review]

**Review by mem — APPROVED** [review]
